pkg/pillar/cmd/zedagent/parseconfig.go

```diff
@@ -39,15 +39,37 @@ const (
 )
 
 // Returns a configProcessingSkipFlag
-func parseConfig(config *zconfig.EdgeDevConfig, getconfigCtx *getconfigContext,
-	usingSaved bool) bool {
+func parseConfig(getconfigCtx *getconfigContext, config *zconfig.EdgeDevConfig,
+	source configSource) bool {
 
 	// Do not accept new commands from Local profile server while new config
 	// from the controller is being applied.
-	getconfigCtx.localCommands.Lock()
-	defer getconfigCtx.localCommands.Unlock()
+	if getconfigCtx.localCommands != nil {
+		getconfigCtx.localCommands.Lock()
+		defer getconfigCtx.localCommands.Unlock()
+	}
+
+	// Make sure we do not accidentally revert to an older configuration.
+	// This depends on the controller attaching config timestamp.
+	// If not provided, the check is skipped.
+	if config.ConfigTimestamp.IsValid() {
+		configTimestamp := config.ConfigTimestamp.AsTime()
+		if getconfigCtx.lastConfigTimestamp.After(configTimestamp) {
+			log.Warnf("Skipping obsolete device configuration "+
+				"(source: %v, timestamp: %v, currently applied: %v)",
+				source, configTimestamp, getconfigCtx.lastConfigTimestamp)
+			return false
```

Comment:
I don't know if returning true here (indicating it was skipped) is a good idea; skip due to old timestamp vs. skipped due to  in-progress reboot are two different things. Maybe introduce an enum for the return so the caller can tell the difference and log appropriately?

Reply:
Added `configProcessingRetval` enum, now used by `getLatestConfig`, `inhaleDeviceConfig` and `parseConfig`.